app/src/main/java/com/nitramite/courier/UPSStrategy.java

```diff
@@ -45,21 +51,35 @@ public ParcelObject execute(String parcelCode) {
         ParcelObject parcelObject = new ParcelObject(parcelCode);
         ArrayList<EventObject> eventObjects = new ArrayList<>();
         try {
+            UpsTokenPair tokenPair = getXSRFToken();
+            if (tokenPair.getX_CSRF_TOKEN() == null || tokenPair.getX_XSRF_TOKEN_ST() == null) {
+                throw new ParseException("Invalid X-XSRF-TOKEN, got null", 0);
+            }
+
             OkHttpClient client = new OkHttpClient();
 
-            final String json = "{\"Locale\":\"fi_FI\",\"TrackingNumber\":[\"" + parcelCode + "\"]}"; // TODO: make proper way
-            Log.i(TAG, json);
-            RequestBody body = RequestBody.create(json, JSON);
+            JSONArray trackingNumbers = new JSONArray();
+            trackingNumbers.put(parcelCode);
+            JSONObject requestBody = new JSONObject();
+            requestBody.put("Locale", "fi_FI");
```

Comment:
Double checking - I assume this is matched with `loc=fi_FI` in `url ` and not `loc=en_US` in `tokenUrl`?

Reply:
Aaah yeah that is something I didn't check for but in this case getting token with en_US didn't change outcome